operations/operations.py

```diff
@@ -0,0 +1,63 @@
+from functools import partial
+from multiprocessing import Pool
+from typing import Dict, List
+
+import scipp as sc
+
+
+def integrate(dataset: sc.Dataset):
+    integrated = sc.Dataset()
+    for dim in dataset.coords.keys():
+        integrated.coords[dim] = dataset.coords[dim]
+
+    for k in dataset.keys():
+        integrated[k] = sc.sum(dataset[k], "tof")
+
+    return integrated
+
+
+
+def _calc_adj_spectra(center_spec_num, bank_width, num_spectra):
+    col_positions = [-1, 0, 1]  # Take relative positions of col
+
+    # We calculate along a row at a time, rather than a col at a time
+    top_row_center = (center_spec_num - bank_width)
+    bottom_row_center = (center_spec_num + bank_width)
+
+    neighbour_spec = []
+
+    # Top / Middle / Bottom row
+    neighbour_spec.extend(i + top_row_center for i in col_positions)
+    neighbour_spec.extend(i + center_spec_num for i in col_positions)
+    neighbour_spec.extend(i + bottom_row_center for i in col_positions)
+
+    def is_valid(spec_num):
+        # Do not include OOB and the centre spectra
+        return 0 <= spec_num < num_spectra and spec_num != center_spec_num
+
+    neighbour_spec = [i for i in neighbour_spec if is_valid(i)]
+    return neighbour_spec
+
+
+def generate_neighbouring_spectra(num_spectra: int,
+                                  bank_width: int) -> Dict[int, List[int]]:
+    # Generates neighbouring spectra, for example selecting
+    # 5 in the following bank will return a dict with [1,2,3,4,6,7,8,9] where
+    # 1,  2,  3
+    # 3, [5], 6
+    # 7,  8,  9
+
+    # Using multiprocessing as this can be slow for 10k spectra, but is
+    # embarrassingly parallel
+    with Pool() as p:
+        results = p.map(partial(_calc_adj_spectra,
+                                bank_width=bank_width,
+                                num_spectra=num_spectra),
+                        range(num_spectra))
+    assert len(results) == num_spectra
+
+    packed_pixels = {}
+    for i, adj_pixels in enumerate(results):
+        packed_pixels[i] = adj_pixels
+
+    return packed_pixels
```

Comment:
No!

```python
def shift(var, dim, forward, out_of_bounds):
  fill = var[dim, 0].copy()
  fill.values = out_of_bounds
  if forward == True:
    return sc.concatenate(fill, var[dim, :-1], dim)
  else:
    return sc.concatenate(var[dim, 1:], fill, dim)

mask = sc.reshape(spectrum_mask, dims=['y', 'x'], shape=[bank_width, bank_width])
flip = sc.Variable(dims=['neighbor', 'y', 'x'], shape=[8, bank_width, bank_width], dtype=sc.dtype.bool)
fill = True
flip['neighbor', 0] = shift(mask, x, True, fill)
flip['neighbor', 1] = shift(mask, x, False, fill)
flip['neighbor', 2] = shift(mask, y, True, fill)
flip['neighbor', 3] = shift(mask, y, False, fill)
flip['neighbor', 4:6] = shift(flip['neighbor', 0:2], y, True, fill)
flip['neighbor', 6:8] = shift(flip['neighbor', 0:2], y, False, fill)
mask = mask ^ sc.all(flip, 'neighbor')
mask = mask ^ ~sc.any(flip, 'neighbor') # if different out-of-bounds handling needed, regenerate `flip` first
# reshape mask back to 1D (or, recommended: work in XY from the start)
```
This also replaces most of the code you have in `mask_non_sample_regions`.

May need debugging, I wrote this blind.

Reply:
I'm going to push back slightly on this, I knew it would be the controversial one haha

I agree the original approach of generating adj. spectra is not the correct way with scipp to go about this, however I wanted to unblock this review (which has been really useful)

As this will be passed back to a user I also think this code is too complex for any reasonable user; it uses shifts of the underlying data structure with bitwise mask combinations that I genuinely don't believe any scientists should be expected to support. It's just going to end up rewritten or treated as something they hope works, which is one of the grievances I get when a user passes "simple code" to a dev then gets something back they can't understand.

I'm planning on going back and adding X Y support from the start, since it will remove several places where we have to account for bank width....etc. now we have validated the data. But I'd still like to find something much more readable for users. 

Numpy has something that will do this to some extent, but I have to reach for `.values` which is a sign I'm doing something wrong:

`views = numpy.lib.stride_tricks.as_strided(var.values, shape=(3, 3), strides=bank_width)`

Which generates n number of views (that can be iterated) with the correct stride and shape of each. This is a computational pessimism as we need to iterate through views still, but CPU time is cheap, and this is going to be demo'd for buy-in that the framework make things simpler and easier to read

Reply:
> I also think this code is too complex for any reasonable user
I would argue the same about the original code. It is 3x as long and you even need to use multi-threading. Also, I gather that this is slow? How slow? Can you compare to my snippet? Users usually care surpisingly much about waiting time.

Reply:
I agree completely, I'm not advocating for what we currently have it's slow and non-obvious for an end user especially the multi processing part

Using `%time` on a fresh start each time (so there is no caching of results):
Above implementation: 590ms to gen adj spectra, + ~120-150ms per data array = ~1 second

I went to benchmark the above method but it seems to segfault? I can't see anything wrong with this example (I've tweaked it to be standalone):

``` python
import scipp as sc
import numpy as np

def get_adj_pixels(mask):
    def shift(var, dim, forward, out_of_bounds):
      fill = var[dim, 0].copy()
      fill.values = out_of_bounds
      if forward == True:
        return sc.concatenate(fill, var[dim, :-1], dim)
      else:
        return sc.concatenate(var[dim, 1:], fill, dim)

    bank_width = 324

    flip = sc.Variable(dims=['neighbor', 'y', 'x'], shape=[8, bank_width, bank_width], dtype=sc.dtype.bool)
    fill = True
    flip['neighbor', 0] = shift(mask, "x", True, fill)
    flip['neighbor', 1] = shift(mask, "x", False, fill)
    flip['neighbor', 2] = shift(mask, "y", True, fill)
    flip['neighbor', 3] = shift(mask, "y", False, fill)
    flip['neighbor', 4:6] = shift(flip['neighbor', 0:2], "y", True, fill)
    flip['neighbor', 6:8] = shift(flip['neighbor', 0:2], "y", False, fill)
    mask = mask ^ sc.all(flip, 'neighbor')
    return mask ^ ~sc.any(flip, 'neighbor') # if different out-of-bounds handling needed, regenerate `flip` first

mask = sc.Variable(["y", "x"], values=np.full(shape=(324, 324), fill_value=False))
%time get_adj_pixels(mask)
```

We could document each step talking user through what each bit does, mainly how XOR is used to select which pixels to mask with some sort of ASCII diagram, but I have a deep concern about writing something that isn't user maintainable.

I'm just going to ping @OwenArnold since he was originally porting this notebook for the user, so might have a better idea on we where we should strike between simplicity and speed? If he doesn't think there will be an issue for this user then I'm happy to drop my concerns on this one and use the above.

Reply:
Ouch... we hit some bug with special handling when assigning `True` or `False` to `values` (I'll open an issue), do this instead:

```python
    def shift(var, dim, forward, out_of_bounds):
      fill = var[dim, 0:1].copy()
      fill.values = np.full_like(fill.values, out_of_bounds)
      # unchange below
```

```
%time 
x = get_adj_pixels(mask)
CPU times: user 4 µs, sys: 2 µs, total: 6 µs
Wall time: 13.6 µs
```

Reply:
This method wasn't working as we ended up getting a mask which looked like

```
|F|F|F|F|F
|F|T|T|T|F
|F|T|T|T|F
|F|F|F|F|F
```

When I needed basically a sliding view of each 3 by 3 for example:
```
1 | 2 | 3 | 4
5 | 6 | 7 | 8
9 | 1 | 2 | 3
```

Would return
```
1 | 2
5 | 6

1 | 2 | 3
5 | 6 | 7
```

However after running into OOM problems trying to create way to many views I stumbled on scipy Convolution2d which we can use to get the same effect

Whilst writing this I realised there is a bug, our fill value is 1. which means boundary points which get forced to `True` will work, but we need to do the convolution again for `False` with a fill value of 0. so I'll fix that later

Reply:
Fixed version in pair programming:
```python
def get_adj_pixels(mask):
    def shift(var, dim, forward, out_of_bounds):
      fill = var[dim, 0:1].copy()
      fill.values = np.full_like(fill.values, out_of_bounds)
      if forward == True:
        return sc.concatenate(fill, var[dim, :-1], dim)
      else:
        return sc.concatenate(var[dim, 1:], fill, dim)

    bank_width = 324

    def make_flip(fill):
        flip = sc.Variable(dims=['neighbor', 'y', 'x'], shape=[8, ]+mask.shape, dtype=sc.dtype.bool)
        flip['neighbor', 0] = shift(mask, "x", True, fill)
        flip['neighbor', 1] = shift(mask, "x", False, fill)
        flip['neighbor', 2] = shift(mask, "y", True, fill)
        flip['neighbor', 3] = shift(mask, "y", False, fill)
        flip['neighbor', 4:6] = shift(flip['neighbor', 0:2], "y", True, fill)
        flip['neighbor', 6:8] = shift(flip['neighbor', 0:2], "y", False, fill)
        return flip
    mask = mask | sc.all(make_flip(True), 'neighbor') # mask if all neighbors masked
    return mask & sc.any(make_flip(False), 'neighbor') # unmask if no neighbor masked
```